src/diffusers/models/dualencoder_gfn.py

```diff
@@ -0,0 +1,695 @@
+    N = pos.size(0)
+
+    bgraph_adj = torch.sparse.LongTensor(edge_index, edge_type, torch.Size([N, N]))
+
+    if is_sidechain is None:
+        rgraph_edge_index = radius_graph(pos, r=cutoff, batch=batch)  # (2, E_r)
+    else:
+        # fetch sidechain and its batch index
+        is_sidechain = is_sidechain.bool()
+        dummy_index = torch.arange(pos.size(0), device=pos.device)
+        sidechain_pos = pos[is_sidechain]
+        sidechain_index = dummy_index[is_sidechain]
+        sidechain_batch = batch[is_sidechain]
+
+        assign_index = radius(x=pos, y=sidechain_pos, r=cutoff, batch_x=batch, batch_y=sidechain_batch)
+        r_edge_index_x = assign_index[1]
+        r_edge_index_y = assign_index[0]
+        r_edge_index_y = sidechain_index[r_edge_index_y]
+
+        rgraph_edge_index1 = torch.stack((r_edge_index_x, r_edge_index_y))  # (2, E)
+        rgraph_edge_index2 = torch.stack((r_edge_index_y, r_edge_index_x))  # (2, E)
+        rgraph_edge_index = torch.cat((rgraph_edge_index1, rgraph_edge_index2), dim=-1)  # (2, 2E)
+        # delete self loop
+        rgraph_edge_index = rgraph_edge_index[:, (rgraph_edge_index[0] != rgraph_edge_index[1])]
+
+    rgraph_adj = torch.sparse.LongTensor(
+        rgraph_edge_index,
+        torch.ones(rgraph_edge_index.size(1)).long().to(pos.device) * unspecified_type_number,
+        torch.Size([N, N]),
+    )
+
+    composed_adj = (bgraph_adj + rgraph_adj).coalesce()  # Sparse (N, N, T)
+    # edge_index = composed_adj.indices()
+    # dist = (pos[edge_index[0]] - pos[edge_index[1]]).norm(dim=-1)
+
+    new_edge_index = composed_adj.indices()
+    new_edge_type = composed_adj.values().long()
+
+    return new_edge_index, new_edge_type
+
+
+def extend_graph_order_radius(
+    num_nodes,
+    pos,
+    edge_index,
+    edge_type,
+    batch,
+    order=3,
+    cutoff=10.0,
+    extend_order=True,
+    extend_radius=True,
+    is_sidechain=None,
+):
+    if extend_order:
+        edge_index, edge_type = _extend_graph_order(
+            num_nodes=num_nodes, edge_index=edge_index, edge_type=edge_type, order=order
+        )
+        # edge_index_order = edge_index
+        # edge_type_order = edge_type
+
+    if extend_radius:
+        edge_index, edge_type = _extend_to_radius_graph(
+            pos=pos, edge_index=edge_index, edge_type=edge_type, cutoff=cutoff, batch=batch, is_sidechain=is_sidechain
+        )
+
+    return edge_index, edge_type
+
+
+def get_distance(pos, edge_index):
+    return (pos[edge_index[0]] - pos[edge_index[1]]).norm(dim=-1)
+
+
+def eq_transform(score_d, pos, edge_index, edge_length):
+    N = pos.size(0)
+    dd_dr = (1.0 / edge_length) * (pos[edge_index[0]] - pos[edge_index[1]])  # (E, 3)
+    score_pos = scatter_add(dd_dr * score_d, edge_index[0], dim=0, dim_size=N) + scatter_add(
+        -dd_dr * score_d, edge_index[1], dim=0, dim_size=N
+    )  # (N, 3)
+    return score_pos
+
+
+class DualEncoderEpsNetwork(ModelMixin, ConfigMixin):
```

Comment:
Since this is the first graph network NN in this library, we should be extra careful with the naming. 
Is this a universally understandable name? Do you think other graph networks would also use this architecuter? Should we make the name more generic in this case? Can we link to a paper here that defined that model architecture?

Reply:
Link to original code is above now too, I followed up with those authors asking if their's was original. https://github.com/DeepGraphLearning/ConfGF